src/components/KTableData/KTableData.vue

```diff
@@ -0,0 +1,706 @@
+        sortColumnOrder: sortColumnOrder.value,
+        data: tableData.value,
+      })
+    } else {
+      defaultClientSideSorter(columnKey, prevKey, sortColumnOrder.value, tableData.value)
+    }
+  } else if (!props.paginationAttributes?.offset) {
+    debouncedRevalidate()
+  }
+
+  // Emit an event whenever one of the tablePreferences are updated
+  emitTablePreferences()
+}
+
+const pageChangeHandler = ({ page: newPage }: PageChangeData) => {
+  page.value = newPage
+}
+
+const pageSizeChangeHandler = ({ pageSize: newPageSize }: PageSizeChangeData) => {
+  offsets.value = [null]
+  offset.value = null
+  pageSize.value = newPageSize
+  page.value = 1
+
+  // Emit an event whenever one of the tablePreferences are updated
+  emitTablePreferences()
+}
+
+const tableViewPreferences = ref<TablePreferences>({})
+const tableDataPreferences = computed((): TablePreferences => ({ pageSize: pageSize.value, ...tableViewPreferences.value }))
+
+const emitTablePreferences = (): void => {
+  if (tableState.value === 'success') {
+    emit('update:table-preferences', tableDataPreferences.value)
+  }
+}
+
+const getNextOffsetHandler = (): void => {
+  page.value++
+  offset.value = nextOffset.value
+}
+
+const getPreviousOffsetHandler = (): void => {
+  page.value--
+  offset.value = previousOffset.value
+}
+
+// fetcher must be defined, hidePagination must be false
+// if using standard pagination with hidePaginationWhenOptional
+//  - hide if total <= min pagesize
+// if using offset-based pagination with hidePaginationWhenOptional
+//  - hide if neither previous/next offset exists and current data set count is < min pagesize
+const showPagination = computed((): boolean => {
+  return !!(props.fetcher && !props.hidePagination &&
+        !(!props.paginationAttributes?.offset && props.hidePaginationWhenOptional && props.paginationAttributes?.pageSizes && total.value <= props.paginationAttributes?.pageSizes[0]) &&
+        !(props.paginationAttributes?.offset && props.hidePaginationWhenOptional && !previousOffset.value && !nextOffset.value && props.paginationAttributes?.pageSizes && tableData.value.length < props.paginationAttributes?.pageSizes[0]))
+})
+
+watch(fetcherData, (fetchedData: any) => {
+  if (fetchedData?.length && !tableData.value.length) {
+    tableData.value = fetchedData
+  }
+}, { deep: true, immediate: true })
+
+// we want to tie loader to 'pending' since 'validating' is triggered even when pulling from cache, which should result in no loader
+// however, if this is a manual revalidation (triggered by page change, query, etc), display loader when validating
+watch(state, () => {
+  switch (state.value) {
+    case swrvState.PENDING:
+      isTableLoading.value = true
+      break
+    case swrvState.VALIDATING_HAS_DATA:
+      isTableLoading.value = isRevalidating.value
+      break
+    default:
+      isTableLoading.value = false
+      break
+  }
+}, { immediate: true })
+
+watch([stateData, tableState], (newData) => {
```

Comment:
nit: can you rename `newData` to something more descriptive?

Also, it would be better to extract the variables rather than referencing below by index:

```ts
const [newStateData, newTableState] = newData
```

Reply:
Done